workflow/controller/dag.go

```diff
@@ -336,6 +343,10 @@ func (woc *wfOperationCtx) executeDAGTask(dagCtx *dagContext, taskName string) {
 	dependenciesCompleted := true
 	dependenciesSuccessful := true
 	nodeName := dagCtx.taskNodeName(taskName)
+	// Ensure that the generated taskNodeName can be reversed into the original (not expanded) task name
+	if dagCtx.taskNameFromNodeName(nodeName) != task.Name {
+		panic("unreachable: node name cannot be reversed into task name; please file a bug on GitHub")
```

Comment:
I've made sure that this code is unreachable now, but if the naming convention is changed in the future this is meant to catch the discrepancy before the code can be shipped.

Reply:
Also, panics are recovered in the operator, so they fail gracefully: https://github.com/argoproj/argo/blob/c6ef1ff19e1c3f74b4ef146be37e74bd0b748cd7/workflow/controller/operator.go#L159-L168

Reply:
you should panic if there is a nasty bug - we log and report on panics